src/vs/base/node/powershell.ts

```diff
@@ -0,0 +1,331 @@
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as fs from 'fs';
+import * as os from 'os';
+import * as path from 'vs/base/common/path';
+import { env } from 'vs/base/common/process';
+
+const WindowsPowerShell64BitLabel = 'Windows PowerShell (x64)';
+const WindowsPowerShell32BitLabel = 'Windows PowerShell (x86)';
+
+// This is required, since parseInt("7-preview") will return 7.
+const IntRegex: RegExp = /^\d+$/;
+
+const PwshMsixRegex: RegExp = /^Microsoft.PowerShell_.*/;
+const PwshPreviewMsixRegex: RegExp = /^Microsoft.PowerShellPreview_.*/;
+
+// The platform details descriptor for the platform we're on
+const isProcess64Bit: boolean = process.arch === 'x64';
+const isOS64Bit: boolean = isProcess64Bit || env.hasOwnProperty('PROCESSOR_ARCHITEW6432');
+
+interface IPowerShellExeDetails {
+	readonly displayName: string;
+	readonly exePath: string;
+}
+
+interface IPossiblePowerShellExe extends IPowerShellExeDetails {
+	exists(): boolean;
+}
+
+class PossiblePowerShellExe implements IPossiblePowerShellExe {
+	public readonly exePath: string;
+	public readonly displayName: string;
+
+	private knownToExist: boolean | undefined;
+
+	constructor(
+		pathToExe: string,
+		installationName: string,
+		{ knownToExist = false }: { knownToExist?: boolean } = {}) {
+
+		this.exePath = pathToExe;
+		this.displayName = installationName;
+		this.knownToExist = knownToExist || undefined;
+	}
+
+	public exists(): boolean {
+		if (this.knownToExist === undefined) {
+			this.knownToExist = fs.existsSync(this.exePath);
+		}
+		return this.knownToExist;
+	}
+}
+
+function getProgramFilesPath(
+	{ useAlternateBitness = false }: { useAlternateBitness?: boolean } = {}): string | null {
+
+	if (!useAlternateBitness) {
+		// Just use the native system bitness
+		return env.ProgramFiles!;
+	}
+
+	// We might be a 64-bit process looking for 32-bit program files
+	if (isProcess64Bit) {
+		return env['ProgramFiles(x86)']!;
+	}
+
+	// We might be a 32-bit process looking for 64-bit program files
+	if (isOS64Bit) {
+		return env.ProgramW6432!;
+	}
+
+	// We're a 32-bit process on 32-bit Windows, there is no other Program Files dir
+	return null;
+}
+
+function getSystem32Path({ useAlternateBitness = false }: { useAlternateBitness?: boolean } = {}): string {
+	const windir: string = env.windir!;
+
+	if (!useAlternateBitness) {
+		// Just use the native system bitness
+		return path.join(windir, 'System32');
+	}
+
+	// We might be a 64-bit process looking for 32-bit system32
+	if (isProcess64Bit) {
+		return path.join(windir, 'SysWOW64');
+	}
+
+	// We might be a 32-bit process looking for 64-bit system32
+	if (isOS64Bit) {
+		return path.join(windir, 'Sysnative');
+	}
+
+	// We're on a 32-bit Windows, so no alternate bitness
+	return path.join(windir, 'System32');
+}
+
+function findPSCoreWindowsInstallation(
```

Comment:
Can this function be broken up as it's quite large?

Reply:
I tried to condense it a bit so it's not as long but I think the code reads well and this file is organized such that each type of PowerShell has its own function so I don't think it needs to be broken down. If more folks think I should do this, I'm happy to be out voted.